ledger/accountdb.go

```diff
@@ -2573,6 +2573,175 @@ func makeOrderedAccountsIter(tx *sql.Tx, accountCount int) *orderedAccountsIter
+		}
+
+		if len(addrbuf) != len(addr) {
+			err = fmt.Errorf("account DB address length mismatch: %d != %d", len(addrbuf), len(addr))
+			return 0, err
+		}
+
+		copy(addr[:], addrbuf)
+
+		var accountData baseAccountData
+		err = protocol.Decode(buf, &accountData)
+		if err != nil {
+			return 0, err
+		}
+		err = baseCb(addr, &accountData, buf)
+		if err != nil {
+			return 0, err
+		}
+
+		err = processAllResources(resRows, addr, &accountData, rowid, resCb)
+		if err != nil {
+			return 0, err
+		}
+
+		count++
+		if accountCount > 0 && count == accountCount {
+			// we're done with this iteration.
+			return count, nil
+		}
+	}
+
+	return count, nil
+}
+
+// LoadAllFullAccounts loads all accounts from balancesTable and resourcesTable.
+// On every account full load it invokes acctCb callback to report progress and data.
+func LoadAllFullAccounts(
+	ctx context.Context, tx *sql.Tx,
+	balancesTable string, resourcesTable string,
+	acctCb func(basics.Address, basics.AccountData),
+) (count int, err error) {
+	baseRows, err := tx.QueryContext(ctx, fmt.Sprintf("SELECT rowid, address, data FROM %s ORDER BY rowid", balancesTable))
+	if err != nil {
+		return
+	}
+	// iterate over the existing resources
+	resRows, err := tx.QueryContext(ctx, fmt.Sprintf("SELECT addrid, aidx, rtype, data FROM %s ORDER BY addrid, aidx, rtype", resourcesTable))
+	if err != nil {
+		return
+	}
+
+	var address basics.Address
+	var baseAcct baseAccountData
+	var ad basics.AccountData
+	var resCounter uint32 = 0
+
+	baseCb := func(addr basics.Address, accountData *baseAccountData, encodedAccountData []byte) (err error) {
+		baseAcct = *accountData
+		ad = baseAcct.GetAccountData()
+		copy(address[:], addr[:])
+		return nil
+	}
+
+	resCb := func(addr basics.Address, cidx basics.CreatableIndex, ctype basics.CreatableType, resData *resourcesData, encodedResourceData []byte) error {
+		if resData.IsApp() && resData.IsOwning() {
+			if ad.AppParams == nil {
+				ad.AppParams = make(map[basics.AppIndex]basics.AppParams)
+			}
+			ad.AppParams[basics.AppIndex(cidx)] = resData.GetAppParams()
+			resCounter++
+		}
+		if resData.IsApp() && resData.IsHolding() {
+			if ad.AppLocalStates == nil {
+				ad.AppLocalStates = make(map[basics.AppIndex]basics.AppLocalState)
+			}
+			ad.AppLocalStates[basics.AppIndex(cidx)] = resData.GetAppLocalState()
+			resCounter++
+		}
+
+		if resData.IsAsset() && resData.IsOwning() {
+			if ad.AssetParams == nil {
+				ad.AssetParams = make(map[basics.AssetIndex]basics.AssetParams)
+			}
+			ad.AssetParams[basics.AssetIndex(cidx)] = resData.GetAssetParams()
+			resCounter++
+		}
+		if resData.IsAsset() && resData.IsHolding() {
+			if ad.Assets == nil {
+				ad.Assets = make(map[basics.AssetIndex]basics.AssetHolding)
+			}
+			ad.Assets[basics.AssetIndex(cidx)] = resData.GetAssetHolding()
+			resCounter++
+		}
+
+		return nil
+	}
+
+	if baseAcct.TotalAppParams+baseAcct.TotalAppLocalStates+baseAcct.TotalAssetParams+baseAcct.TotalAssets == resCounter {
```

Comment:
I can't see where this variable is being populated.

Reply:
`baseCb` sets `baseAcct` variable, and `resCb` updates the `resCounter` counter

Reply:
yes, but the `baseAcct` is called only via `processAllBaseAccountRecords`, and the above comparison is made prior to that.

Reply:
oh, right, good catch. fixed.